### PR TITLE
Added tests for importing perfect excel feature

### PR DESCRIPTION
**Pull request by IamEzio**
Fixes part of #3027 

This PR adds tests for importing perfect Excel file. Perfect excel file means an excel file with a single sheet and a single table.

**Screenshots**

NA

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by IamEzio** [comment]
@dmos62 @Anish9901 PTAL. Thanks!

**Review by dmos62 — COMMENTED** [review]
Tough to review while #3059 is not merged (can't differentiate the changes). Can you request my review when the mentioned PR is merged? By the way, I recommend using Github's review requesting feature (top right of the thread interface): makes it easier for us to triage PRs and so on: not a big deal, but can be useful to me sometimes.

**Comment by IamEzio** [comment]
@dmos62 this PR is ready now. PTAL. Thanks!

File: mathesar/tests/api/test_data_file_api.py [hunk 1]
```diff
@@ -124,6 +124,18 @@ def test_data_file_create_json(client, patents_json_filepath, header):
     )
 
 
+@pytest.mark.parametrize('header', [True, False])
+def test_data_file_create_excel(client, patents_excel_filepath, header):
+    num_data_files = DataFile.objects.count()
+
+    with open(patents_excel_filepath, 'rb') as excel_file:
+        data = {'file': excel_file, 'header': header}
```
Comment (dmos62) [inline mathesar/tests/api/test_data_file_api.py:132]: `header` is too vague. Use `has_header` or something along those lines. That will make it clear that this is a boolean flag.

File: mathesar/tests/api/test_table_api.py [hunk 2]
```diff
@@ -44,6 +44,19 @@ def missing_keys_json_data_file():
     return data_file
 
 
+@pytest.fixture
+def patents_excel_data_file():
+    data_filepath = 'mathesar/tests/data/patents.xlsx'
```
Comment (dmos62) [inline mathesar/tests/api/test_table_api.py:49]: Use the `patents_excel_filepath` fixture here.

File: mathesar/tests/imports/test_excel.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    return data_file
+
+
+@pytest.fixture()
+def schema(engine, test_db_model):
+    create_schema(TEST_SCHEMA, engine)
+    schema_oid = get_schema_oid_from_name(TEST_SCHEMA, engine)
+    yield Schema.current_objects.create(oid=schema_oid, database=test_db_model)
+    with engine.begin() as conn:
+        conn.execute(text(f'DROP SCHEMA "{TEST_SCHEMA}" CASCADE;'))
```
Comment (dmos62) [inline mathesar/tests/imports/test_excel.py:28]: This `schema` fixture isn't needed. I'm aware our fixture setup is pretty esoteric.

File: mathesar/tests/api/test_table_api.py [hunk 4]
```diff
@@ -2010,3 +2023,32 @@ def test_create_table_with_nested_json_objects(client, schema):
...
+        "Patent Number",
+        "Application SN",
+        "Title",
+        "Patent Expiration Date",
+    ]
+
+    check_create_table_response(
+        client, table_name, expt_name, patents_excel_data_file, schema, first_row,
+        column_names, import_target_table=None
+    )
```
Comment (dmos62) [inline mathesar/tests/api/test_table_api.py:2054]: Please refactor this and other API tests you wrote (meaning tests within `mathesar/tests/api/`) to import tables via HTTP API calls, not Python calls. Same for querying them. Here, the patents table is created via `DataFile.objects.create` inside the fixture you wrote.

We talked about this before, but the motivation for having HTTP API tests is that we want these tests to integrate as much of the backend stack as possible.

**Review by dmos62 — APPROVED** [review]

**Review by dmos62 — CHANGES_REQUESTED** [review]
Accidentally clicked "Approve". Meant to request changes.

File: mathesar/tests/api/test_data_file_api.py (hunk 1 above)
Comment (IamEzio) [reply mathesar/tests/api/test_data_file_api.py]: Updated it now. 👍🏻

File: mathesar/tests/api/test_table_api.py (hunk 2 above)
Comment (IamEzio) [reply mathesar/tests/api/test_table_api.py]: Done

File: mathesar/tests/imports/test_excel.py (hunk 3 above)
Comment (IamEzio) [reply mathesar/tests/imports/test_excel.py:28]: We need the `schema` fixture because we want to test that we are not able to create duplicate tables with the same name in a schema. 
I tried removing the fixture completely and then tried to create duplicate tables but could not replicate the scenario. If you could provide some more insights on that, I will update it accordingly.

**Review by Anish9901 — CHANGES_REQUESTED** [review]

File: mathesar/tests/imports/test_excel.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    assert table.sa_num_records() == 1393
+
+    with pytest.raises(DuplicateTable):
+        create_table_from_data_file(data_file, table_name, schema)
+
+
+def test_excel_upload_table_imported_to(data_file, schema):
+    table = create_table_from_data_file(data_file, "NASA", schema)
+    data_file.refresh_from_db()
+    assert data_file.table_imported_to == table
```
[the remark this answers is not in the document]
Comment (IamEzio) [reply mathesar/tests/imports/test_excel.py:86]: I beg to differ. I think if we create a table using a datafile object, and once the db is refreshed, that datafile object should have its `table_imported_to` attribute set to that newly created table.

[the remark this answers is not in the document]
Comment (Anish9901) [reply mathesar/tests/imports/test_excel.py:86]: Umm, you are right. Sorry, I got it mixed up with `import_target`.

File: mathesar/tests/imports/test_excel.py (hunk 3 above)
Comment (Anish9901) [reply mathesar/tests/imports/test_excel.py:28]: No you don't, just get the oid of the schema from the name using `get_schema_oid_from_name` and then trigger reflection in the `Schema` model using `objects` instead of `current_objects`

File: mathesar/tests/api/test_table_api.py (hunk 2 above)
Comment (Anish9901) [reply mathesar/tests/api/test_table_api.py]: you can pass in `patents_excel_filepath` as a param to the `patents_excel_data_file`.

Comment (IamEzio) [reply mathesar/tests/api/test_table_api.py]: Done

File: mathesar/tests/imports/test_excel.py (hunk 3 above)
Comment (IamEzio) [reply mathesar/tests/imports/test_excel.py]: Cool. Updated it now. PTAL

**Review by Anish9901 — APPROVED** [review]
Looks good after the changes @IamEzio, Thanks!

**Review by dmos62 — APPROVED** [review]
